src/ModelContextProtocol/Server/McpServer.cs

```diff
@@ -411,10 +436,22 @@ private void SetToolsHandler(McpServerOptions options)
                     }
                     while (nextCursor is not null);
                 }
+                
+                if(progressToken is not null)
+                {
+                    await this.NotifyProgressAsync(new()
+                    {
+                        ProgressToken = progressToken,
+                        Progress = resultTools.Count,
+                        Total = resultTools.Count,
+                        Message = "Finished listing all tools.",
+                    }, cancellationToken).ConfigureAwait(false);
+                }
 
                 return result;
             };
 
+            // TODO: We may want to inject something like an IProgress instance here for optional use in tool handlers. 
```

Comment:
An `IProgress<ProgressNotificationValue>` can already be injected into a tool written with McpServerTool / McpPromptTool (either programmatically or with the attributes). Let's stick with that for now. If we decide we need something more, I'd rather expose a method like `static IProgress<ProgressNotificationValue> GetIProgress(this RequestContext<CallContextParams> context)`, which that injection mechanism could use but which could also be called by a handler if it wanted the same logic.